io/zenoh-links/zenoh-link-shm/src/unicast.rs

```diff
@@ -0,0 +1,439 @@
+        )?;
+
+        // create uplink
+        let mut uplink = PipeR::new(&path_uplink, access_mode).await?;
+
+        // create listening task
+        let listening_task_handle = async_std::task::spawn(async move {
+            // read invitation from uplink pipe and check it's correctness
+            let mut invitation = [0; 4];
+            uplink.read_exact(&mut invitation).await?;
+            if invitation != PIPE_INVITATION {
+                bail!("Unexpected invitation received during pipe handshake!")
+            }
+
+            // create downlink
+            let mut downlink = PipeW::new(&path_downlink).await?;
+            // echo invitation to other party
+            downlink.write_all(&invitation).await?;
+
+            // send newly established link to manager
+            manager
+                .send_async(LinkUnicast(Arc::new(UnicastPipe {
+                    r: async_std::sync::RwLock::new(uplink),
+                    w: async_std::sync::RwLock::new(downlink),
+                    local,
+                    remote,
+                })))
+                .await?;
+
+            ZResult::Ok(())
+        });
+
+        Ok(Self {
+            listening_task_handle,
+            uplink_locator: c_local,
+        })
+    }
+
+    async fn stop_listening(self) {
+        self.listening_task_handle.cancel().await;
+    }
+}
+
+struct UnicastPipeClient;
+impl UnicastPipeClient {
+    async fn connect_to(endpoint: EndPoint) -> ZResult<UnicastPipe> {
+        let (path_uplink, path_downlink, access_mode) = parse_pipe_endpoint(&endpoint);
+
+        // create Locators
+        let local = Locator::new(
+            endpoint.protocol(),
+            path_downlink.as_str(),
+            endpoint.metadata(),
+        )?;
+        let remote = Locator::new(
+            endpoint.protocol(),
+            path_uplink.as_str(),
+            endpoint.metadata(),
+        )?;
+
+        // create uplink
+        let mut uplink = PipeW::new(&path_uplink).await?;
+        // create downlink
+        let mut downlink = PipeR::new(&path_downlink, access_mode).await?;
+
+        // write invitation
+        uplink.write_all(PIPE_INVITATION).await?;
+
+        // read responce
+        let mut responce: [u8; 4] = [0, 0, 0, 0];
+        downlink.read_exact(&mut responce).await?;
+        // check responce
+        if responce != PIPE_INVITATION {
+            bail!("Unexpected responce received during pipe handshake!")
+        }
+
+        Ok(UnicastPipe {
+            r: async_std::sync::RwLock::new(downlink),
+            w: async_std::sync::RwLock::new(uplink),
+            local,
+            remote,
+        })
+    }
+}
+
+struct UnicastPipe {
+    r: async_std::sync::RwLock<PipeR>,
+    w: async_std::sync::RwLock<PipeW>,
+    local: Locator,
+    remote: Locator,
+}
+impl Drop for UnicastPipe {
+    fn drop(&mut self) {}
+}
+#[async_trait]
+impl LinkUnicastTrait for UnicastPipe {
+    async fn close(&self) -> ZResult<()> {
```

Comment:
Close could happen concurrently. In this particular case is a no-op, but in case something needs to be done on the link, a mutex should be used here.